java/src/test/java/vowpalWabbit/ClosingTest.java

```diff
@@ -0,0 +1,214 @@
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.junit.Assert.*;
+
+/**
+ * Test various functions related to closing the model.
+ *
+ * These tests may fail as they are likely related to the speed of the computer on which they are run.
+ * Created by ryan.deak on 7/21/17.
+ */
+@NotThreadSafe
+public class ClosingTest  extends VWTestHelper {
+    private static final int DELTA_TIME_MS = 50; // milliseconds
+    private static final int PASSES = 25000;
+
+    /**
+     * Enable to see diagnostics during tests.
+     */
+    private static final boolean DIAGNOSTICS = false;
+
+    @Test
+    public void testCloseAsyncIsNearImmediate() throws Exception {
+        VWScalarLearner vw = model(PASSES);
+
+        long t1 = System.currentTimeMillis();
+        Callable<Boolean> closer = vw.closer();
+        long t2 = System.currentTimeMillis();
+        closer.call();
+        long t3 = System.currentTimeMillis();
+
+        assertEquals("Should only take a few milliseconds to get future", 0, t2 - t1, DELTA_TIME_MS);
+        assertNotEquals("Should take longer to finish synchronous closing", 0, t3 - t1, DELTA_TIME_MS);
+    }
+
+    @Test
+    public void closeMultipleTimes() throws IOException {
+        try {
+            final VWScalarLearner model = model(1);
+            model.close();
+            model.close();
+        }
+        catch (Exception e) {
+            final PrintWriter str = new PrintWriter(new StringWriter());
+            e.printStackTrace(str);
+            fail("Shouldn't throw exception but threw " + str.toString());
+        }
+    }
+
+    @Test
+    public void closerThenClose() {
+        try {
+            final VWScalarLearner model = model(1);
+            model.closer().call();
+            model.close();
+        }
+        catch (Exception e) {
+            final PrintWriter str = new PrintWriter(new StringWriter());
+            e.printStackTrace(str);
+            fail("Shouldn't throw exception but threw " + str.toString());
+        }
+    }
+
+    @Test
+    public void closeThenCloser() {
+        try {
+            final VWScalarLearner model = model(1);
+            model.close();
+            model.closer().call();
+        }
+        catch (Exception e) {
+            final PrintWriter str = new PrintWriter(new StringWriter());
+            e.printStackTrace(str);
+            fail("Shouldn't throw exception but threw " + str.toString());
+        }
+    }
+
+    @Test
+    public void multipleAsynchronousCloserCallsReturnTrueOnlyOnce() throws Exception {
+        final int numClosers = 10;   // Should be greater than 2 to avoid confusion.
```

Comment:
Shows asynchronous closing multiple times.  This (probabilistically) demonstrates thread-safety and shows only one close has an effect.